<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>

<head>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8">
    <link rel="stylesheet" type="text/css" href="../../_css/labs-sd.css"><!--/link-->

    <link href="../../_prettify/prettify.css" type="text/css" rel="stylesheet" />
    <script type="text/javascript" src="../../_prettify/prettify.js"></script>

    <title>Exceptions</title>

    <meta name="Generator" content="UltraEdit">
    <meta name="Author" content="Naércio Magaia">
    <meta name="Keywords" content="Exception in Java, tutorial">
    <meta name="Description" content="">
</head>

<body onload="prettyPrint()">

<div class="contexto">
    <p>
        <a href="../../index.html">Labs SD</a> &gt;
        <a href="../index.html">Ferramentas</a>
    </p>
</div>

<div class="titulo">
    <h1>Exceções</h1>
</div>

<p>
As exceções são usadas na linguagem Java para assinalar que algo não correu como esperado.<br />
São classes que herdam de <tt>java.lang.Exception</tt> e cujos objetos podem ser atirados (<i>throw</i>) e apanhados (<i>caught</i>).
</p>
<p>
As exceções que herdam de <tt>java.lang.RuntimeException</tt> (RTE) são chamadas exceções não verificadas (<i>unchecked exceptions</i>).
Neste caso, o compilador não obriga o programador a declarar se apanha ou se atira.
Por este motivo, qualquer linha de código pode atirar uma exceção destas.
A mais conhecida é a <tt>NullPointerException</tt> (NPE).
</p>
<p>
As exceções que herdam de <tt>java.lang.Exception</tt> são chamadas exceções verificadas (<i>checked exceptions</i>), no sentido, em que a sua utilização é verificada pelo compilador.
Nestes casos é preciso explicitar se se apanha a exceção (<i>catch</i>) ou se se lança (<i>throws</i>).
Normalmente, se não se vai tentar recuperar a exceção, pode simplesmente dizer que se atira.
É preferível atirar do que fazer um falso tratamento de exceção.
</p>


<p>&nbsp;</p>
<h2>Abordagens ao tratamento de exceções</h2>

<p>
Podemos ter as seguintes abordagens em relação às exceções:
</p>
<ul>
    <li>Deixar-passar (<i>pass-through</i>)</li>
    <li>Apanhar-e-tratar (<i>catch-and-handle</i>)</li>
    <ul>
	    <li><span style="color: red">Apanhar-e-ignorar! (<i>swallow!</i>)</span></li>
        <li>Apanhar-registar-e-atirar (<i>logging</i>)</li>
        <li>Apanhar-embrulhar-e-atirar (<i>wrapping</i>)</li>
        <li>Apanhar-e-recuperar (<i>recovery</i>)</li>
    </ul>
</ul>

<p>
A abordagem 'apanhar-e-ignorar' é claramente errada porque perde informação e torna muito mais difícil diagnosticar e resolver problemas.
</p>


<p>&nbsp;</p>
<h2>Exceções impressas na consola</h2>

<p>
Os <i>outputs</i> seguintes foram produzidos por uma aplicação que usa <i>sockets</i> para comunicação, e ilustram diversas situações.
</p>
<p>

<p>
Na situação abaixo, o servidor tenta criar o <i>socket</i> com um porto fora do intervalo [0;65535].
</p>

<pre class="prettyprint lang-java">
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:293)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Port value out of range: 65536
	at java.net.ServerSocket.&lt;init&gt;(ServerSocket.java:232)
	at java.net.ServerSocket.&lt;init&gt;(ServerSocket.java:128)
	at example.SocketServer.main(SocketServer.java:24)
    ... 6 more

</pre>

<p>
Na situação abaixo, o cliente tenta ligar ao servidor por meio dum porto incorreto e não consegue.
</p>

<pre class="prettyprint lang-java">
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:293)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.net.ConnectException: Connection refused: connect
	at java.net.DualStackPlainSocketImpl.connect0(Native Method)
	at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:79)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:345)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at java.net.Socket.&lt;init&gt;(Socket.java:434)
	at java.net.Socket.&lt;init&gt;(Socket.java:211)
	at example.SocketClient.main(SocketClient.java:32)
    ... 6 more

</pre>

<p>
As linhas mais importantes dos <i>outputs</i> são as começadas por <b><tt>Caused by</tt></b>.
É aí que se podem encontrar as mensagens das exceções.
Dado que uma exceção pode ter outra exceção aninhada (<i>cause</i>) pode ser necessário consultar várias linhas para perceber o que causou a exceção de topo.
</p>
<p>
As linhas começadas por <b><tt>at</tt></b> indicam o contexto de execução.
Observando estas linhas é possível ver o conteúdo da pilha de execução do programa, que diz que parte do código estava a chamar que outra parte.
</p>


<p>&nbsp;</p>
<h2>Código de tratamento de exceções</h2>
        
O seguinte código mostra <span style="color: red"><b>maus exemplos</b></span> de tratamento de exceções.
</p>
<pre class="prettyprint lang-java">
    // ANTI-padrão 'apanhar-e-ignorar!'
    // Ignorar a exceção sem dizer nada a ninguém.
    // Evitar! 
    try {
        doSomething();
        
    } catch(Exception e) {
    }
    
    ...
    
    // ANTI-padrão 'apanhar-imprimir-e-ignorar!' 
    // Imprimir o <i>stack trace</i>, não resolve nada.
    // O programa vai "rebentar" mais à frente, onde será mais difícil perceber porquê.
    // Evitar também! 
    try {
        doSomething();
        
    } catch(Exception e) {
        e.printStackTrace();
    }


</pre>
<p>
Ignorar a exceção torna muito mais difícil detetar e corrigir erros no código!
</p>

<p>&nbsp;</p>

<p>
Vamos então ilustrar alguns <span style="color: green"><b>bons exemplos</b></span>:
</p>
<pre class="prettyprint lang-java">

    // padrão 'deixar-passar'
    // Se a exceção não vai ser tratada, mais vale lançá-la (<i>throws</i>)
    public static void main(String[] args) throws Exception {
        doSomething();
    }

    ...
    
    // padrão 'apanhar-imprimir-e-atirar'
    // Registar onde foi apanhada a exceção, mas voltar a atirá-la para que seja tratada depois
    try {
        doSomething();
    
    } catch(MyException e) {
        System.err.println("Caught exception when doing something: " + e);
        System.err.println("Rethrowing"):
        throw e;
    }
    
    
    // padrão 'apanhar-embrulhar-e-atirar'
    // Apanhar exceção da camada inferior
    // Envolver com mais contexto (novo tipo, mensagem de erro melhor)
    // Atirar
    try {
        doSomething();
    
    } catch(MyLowerLevelException e) {
        System.err.println("Caught exception when doing something: " + e);
        System.err.println("Wrapping and throwing, adding meaningful message")
        throw new MyHigherLevelException("Failed to do something.", e);
    }


</pre>

<p>&nbsp;</p>
<h2>Conclusão</h2>

<p>
Em resumo, um bom tratamento de exceções é muito importante em Sistemas Distribuídos.<br />
As melhores estratégias a seguir são: 'deixar-passar', 'apanhar-*-atirar' e 'apanhar-e-recuperar' quando possível.
</p>


<p>&nbsp;</p>

<hr />

<div class="rodape">
<p>
&copy; Docentes de Sistemas Distribu&iacute;dos,
<a href="http://www.dei.tecnico.ulisboa.pt/">Dep. Eng. Informática</a>,
<a href="http://www.ist.eu">Técnico Lisboa</a><br />
</p>
</div>

</body></html>